### ISSUE TITLE
Phase 51.3: MultiModalFusion — Attention-Based Cross-Modal Fusion & Perceiver Tokenization

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#988]
### Overview
Implement attention-based cross-modal fusion mechanisms for combining information from multiple modalities into unified representations. Uses Perceiver-style cross-attention, gated fusion, and hierarchical multi-modal interaction layers.

### Architecture

```
┌──────────────────────────────────────────────────────────────────┐
│                      MultiModalFusion                            │
│                                                                  │
│  ┌─────────┐ ┌─────────┐ ┌─────────┐ ┌─────────┐              │
│  │  Text   │ │  Image  │ │  Audio  │ │  Video  │   Modality   │
│  │ Tokens  │ │ Patches │ │ Frames  │ │ Clips   │   Inputs     │
│  └────┬────┘ └────┬────┘ └────┬────┘ └────┬────┘              │
│       │           │           │           │                     │
│  ┌────▼───────────▼───────────▼───────────▼────┐               │
│  │         Perceiver Cross-Attention           │               │
│  │   Latent Array × Input Cross-Attention      │               │
│  └─────────────────┬───────────────────────────┘               │
│                    │                                            │
│  ┌─────────────────▼───────────────────────────┐               │
│  │          Gated Fusion Layer                  │               │
│  │  σ(Wg·[h_text; h_image; ...]) ⊙ tanh(Wv·h) │               │
│  └─────────────────┬───────────────────────────┘               │
│                    │                                            │
│  ┌─────────────────▼───────────────────────────┐               │
│  │      Self-Attention Transformer Blocks       │               │
│  │   (Fused multi-modal representation)         │               │
│  └─────────────────┬───────────────────────────┘               │
│                    │                                            │
│  ┌─────────────────▼───────────────────────────┐               │
│  │    Task-Specific Output Heads                │               │
│  │  [Classification] [VQA] [Captioning] [...]   │               │
│  └─────────────────────────────────────────────┘               │
└──────────────────────────────────────────────────────────────────┘
```

### Key Components

1. **PerceiverCrossAttention**: Perceiver-style cross-attention mapping high-dim inputs to fixed-size latent array
2. **GatedFusionLayer**: Gated multi-modal fusion with learned modality importance weights
3. **CrossModalAttention**: Flamingo-style interleaved cross-attention between vision and language streams
4. **BottleneckTokenizer**: Compress variable-length modality inputs into fixed token budgets
5. **ModalityDropout**: Random modality dropout during training for robustness to missing modalities
6. **HierarchicalFusion**: Early, mid, and late fusion strategies with configurable fusion points
7. **AttentionPooling**: Multi-head attention pooling for aggregating token-level to sequence-level representations

### Interface

```typescript
interface IMultiModalFusion {
  fuse(inputs: Map<ModalityType, LatentEmbedding>): Promise<FusedRepresentation>;
  fusePair(a: LatentEmbedding, b: LatentEmbedding): Promise<FusedRepresentation>;
  setFusionStrategy(strategy: FusionStrategy): void;
  getModalityWeights(): Map<ModalityType, number>;
  enableModalityDropout(rate: number): void;
}

interface FusedRepresentation {
  tokens: Float32Array[];           // fused token-level representations
  pooled: Float32Array;             // fused pooled representation
  modalityAttentionWeights: Map<ModalityType, Float32Array>;
  fusionMetadata: FusionMetadata;
}

type FusionStrategy = "perceiver" | "gated" | "cross_attention" | "early" | "late" | "hierarchical";
```

### Implementation Details

- **Perceiver IO Architecture**: Iterative cross-attention with latent bottleneck (default 256 latent tokens)
- **Flamingo-Style Gating**: tanh-gated cross-attention inserted between frozen LM layers
- **Q-Former Bridge**: BLIP-2-style querying transformer bridging vision and language
- **Attention Masking**: Causal and bidirectional masking options for different fusion modes
- **Memory Efficiency**: Flash attention integration for O(N) memory cross-modal attention

### Acceptance Criteria

- [ ] Perceiver cross-attention processes arbitrary-length inputs from 2+ modalities
- [ ] Gated fusion layer learns meaningful modality importance weights
- [ ] Modality dropout enables inference with missing modalities (≤10% performance drop)
- [ ] Cross-modal attention patterns are interpretable and visualizable
- [ ] Hierarchical fusion outperforms early/late fusion baselines
- [ ] Unit tests with ≥90% coverage
- [ ] Memory usage scales linearly with latent array size, not input length

### Dependencies
- Phase 51.1 (ModalityEncoder) — provides per-modality embeddings
- Phase 51.2 (CrossModalAligner) — aligned embedding space
- Phase 28 (Attention & Cognitive Resource Management) — attention mechanisms

### References
- Jaegle et al. (2021) — Perceiver / Perceiver IO
- Alayrac et al. (2022) — Flamingo cross-attention
- Li et al. (2023) — BLIP-2 Q-Former
- Nagrani et al. (2021) — Attention bottlenecks for multi-modal fusion
- Arevalo et al. (2017) — Gated multi-modal units

*Part of [Phase 51 Planning](#985)*
